cirq-core/cirq/ops/control_values.py

```diff
@@ -0,0 +1,223 @@
+
+class ControlValues:
+    def __init__(
+        self, control_values: Sequence[Union[int, Collection[int], Type['ControlValues']]]
+    ):
+        if len(control_values) == 0:
+            self.vals = cast(Tuple[Tuple[int, ...], ...], (()))
+            self.num_variables = 0
+            self.nxt = None
+            self.itr = None
+            return
+        self.itr = None
+        self.nxt = None
+
+        if len(control_values) > 1:
+            self.nxt = ControlValues(control_values[1:])
+
+        if isinstance(control_values[0], ControlValues):
+            aux = control_values[0].copy()
+            aux.And(self.nxt)
+            self.vals, self.num_variables, self.nxt = aux.vals, aux.num_variables, aux.nxt
+            self.vals = cast(Tuple[Tuple[int, ...], ...], self.vals)
+            return
+
+        val = control_values[0]
+        if isinstance(val, int):
+            self.vals = _from_int(val)
+        elif isinstance(val, (list, tuple)):
+            if isinstance(val[0], int):
+                self.vals = _from_sequence_int(val)
+            else:
+                self.vals = _from_sequence_sequence(val)
+        self.num_variables = len(self.vals[0])
+
+    def And(self, other):  # pylint: disable=invalid-name
+        # Cartesian product of all combinations in self x other
+        if other is None:
+            return
+        other = other.copy()
+        cur = self
+        while cur.nxt is not None:
+            cur = cur.nxt
+        cur.nxt = other
+
+    def __call__(self):
+        return self.__iter__()
+
+    def __iter__(self):
+        nxt = self.nxt if self.nxt else lambda: [()]
+        if self.num_variables:
+            self.itr = itertools.product(self.vals, nxt())
+        else:
+            self.itr = itertools.product(*(), nxt())
+        return self.itr
+
+    def copy(self):
+        if self.num_variables == 0:
+            new_copy = ControlValues([])
+        else:
+            new_copy = ControlValues(
+                [
+                    copy.deepcopy(self.vals),
+                ]
+            )
+        new_copy.nxt = None
+        if self.nxt:
+            new_copy.nxt = self.nxt.copy()
+        return new_copy
+
+    def __len__(self):
+        cur = self
+        num_variables = 0
+        while cur is not None:
+            num_variables += cur.num_variables
+            cur = cur.nxt
+        return num_variables
+
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            if key != slice(None, -1, None):
+                raise ValueError('Unsupported slicing')
+            return self.copy().pop()
+        key = int(key)
+        num_variables = len(self)
+        if not 0 <= key < num_variables:
+            key = key % num_variables
+        cur = self
+        while cur.num_variables <= key:
+            key -= cur.num_variables
+            cur = cur.nxt
+        return cur
+
+    def __eq__(self, other):
+        if not isinstance(other, ControlValues):
```

Comment:
I guess the intention here is to make this equal to other possible ways to build a control value?  If so I think you want to check the types more and if it doesn't equal any of those types, return `NotImplemented` (python's equals contract)

Reply:
done